src/xenia/cpu/xex_module.h

```diff
@@ -30,61 +29,158 @@ class Runtime;
-  static bool GetOptHeader(const xex2_header* header, xe_xex2_header_keys key,
+  static bool GetOptHeader(const xex2_header* header, xex2_header_keys key,
                            void** out_ptr);
-  bool GetOptHeader(xe_xex2_header_keys key, void** out_ptr) const;
+  bool GetOptHeader(xex2_header_keys key, void** out_ptr) const;
 
   // Ultra-cool templated version
   // Special case: if key & 0xFF == 0x00, this function will return the value,
   // not a pointer! This assumes out_ptr points to uint32_t.
   template <typename T>
-  static bool GetOptHeader(const xex2_header* header, xe_xex2_header_keys key,
+  static bool GetOptHeader(const xex2_header* header, xex2_header_keys key,
                            T* out_ptr) {
     return GetOptHeader(header, key, reinterpret_cast<void**>(out_ptr));
   }
 
   template <typename T>
-  bool GetOptHeader(xe_xex2_header_keys key, T* out_ptr) const {
+  bool GetOptHeader(xex2_header_keys key, T* out_ptr) const {
     return GetOptHeader(key, reinterpret_cast<void**>(out_ptr));
   }
 
   static const xex2_security_info* GetSecurityInfo(const xex2_header* header);
 
+  const PESection* GetPESection(const char* name);
+
   uint32_t GetProcAddress(uint16_t ordinal) const;
   uint32_t GetProcAddress(const char* name) const;
 
-  bool ApplyPatch(XexModule* module);
   bool Load(const std::string& name, const std::string& path,
-            const void* xex_addr, size_t xex_length);
-  bool Load(const std::string& name, const std::string& path, xe_xex2_ref xex);
+            const void* xex_addr, size_t xex_length,
+            XexModule* base_xex = nullptr);
+  bool LoadContinue();
   bool Unload();
 
   const std::string& name() const override { return name_; }
   bool is_executable() const override {
     return (xex_header()->module_flags & XEX_MODULE_TITLE) != 0;
   }
 
+  bool is_valid_executable() const {
+    assert_not_zero(base_address_);
+    if (!base_address_) {
+      return false;
+    }
+    uint8_t* buffer = memory()->TranslateVirtual(base_address_);
+    return *(uint32_t*)buffer == 0x905A4D;
+  }
+
+  bool is_patch() const {
+    assert_not_null(xex_header());
+    if (!xex_header()) {
+      return false;
+    }
+    return (xex_header()->module_flags &
+            (XEX_MODULE_MODULE_PATCH | XEX_MODULE_PATCH_DELTA |
+             XEX_MODULE_PATCH_FULL));
+  }
+
   bool ContainsAddress(uint32_t address) override;
 
+  static void DecryptBuffer(const uint8_t* session_key,
+                            const uint8_t* input_buffer,
+                            const size_t input_size, uint8_t* output_buffer,
+                            const size_t output_size);
+
+  uint8_t* HostData() {
+    if (base_address_)
+      return memory()->TranslateVirtual(base_address_);
+    else
+      return nullptr;
+  }
+
  protected:
   std::unique_ptr<Function> CreateFunction(uint32_t address) override;
 
+  void DecryptSessionKey(const uint8_t* key);
```

Comment:
Instead of having a function dedicated to doing this - we can do DecryptBuffer instead.
i.e:
```
      DecryptBuffer(xe_xex2_retail_key,
                    reinterpret_cast<const uint8_t*>(sec_header->aes_key), 16,
                    session_key_, 16);
```

Reply:
Good idea, I think there's another part in ReadImageDeltaCompressed that calls rijndael stuff directly which can use that too.